core/trino-main/src/main/java/io/trino/operator/aggregation/GroupedAggregator.java

```diff
@@ -68,10 +75,52 @@ public void processPage(GroupByIdBlock groupIds, Page page)
     {
         if (step.isInputRaw()) {
             accumulator.addInput(groupIds, page.getColumns(inputChannels), getMaskBlock(page));
+            return;
         }
-        else {
-            accumulator.addIntermediate(groupIds, page.getBlock(inputChannels[0]));
+
+        if (rawInputMaskChannel.isEmpty()) {
+            // process partially aggregated data
+            accumulator.addIntermediate(groupIds, page.getBlock(intermediateStateChannel));
+            return;
+        }
+        Block rawInputMaskBlock = page.getBlock(rawInputMaskChannel.getAsInt());
+        if (rawInputMaskBlock instanceof RunLengthEncodedBlock) {
+            if (rawInputMaskBlock.isNull(0)) {
+                // process partially aggregated data
+                accumulator.addIntermediate(groupIds, page.getBlock(intermediateStateChannel));
+            }
+            else {
+                // process raw data
+                accumulator.addInput(groupIds, page.getColumns(inputChannels), getMaskBlock(page));
+            }
+            return;
         }
+
+        // rawInputMaskBlock has potentially mixed partially aggregated and raw data
+        Optional<Block> maskBlock = Optional.of((getMaskBlock(page).map(mask -> andMasks(mask, rawInputMaskBlock)).orElse(rawInputMaskBlock)));
```

Comment:
`maskBlock` shouldn''t be optional since `rawInputMaskBlock` must be present:
```
getMaskBlock(page)
  .map(mask -> andMasks(mask, rawInputMaskBlock))
  .orElse(rawInputMaskBlock);
```

Reply:
maskBlock was `Optional` because `addInput` requires it but it is better for var to be not optional so I moved the optional to the next line.